### ISSUE TITLE
[Feature request] Allow exceptions in load_state_dict

### ISSUE THREAD

**Issue by karandwivedi42** [pytorch/pytorch#1990]
Currently, [`load_state_dict`](https://github.com/pytorch/pytorch/blob/2c038f20748c588c9e98f3d549036237ab280ff7/torch/nn/modules/module.py#L337) works only when the loaded state_dict and the Module's state_dict match exactly. 

Problem: For the purpose of fine-tuning or for any slight change in the network, it is not straightforward to load the base layers, leaving one or two unchanged. Either we have to define the old network, load the state and then modify the network, or essentially rewrite the `load_state_dict` function.

Proposed Solution:

```python
    def load_state_dict(self, state_dict, ignored_keys=None):
```

Can I create a PR for the same?

**Comment by apaszke** [comment]
How would `ingored_keys` work? A workaround I'd recomend for partial updates is this:
```python
new_params = model.state_dict()
new_params.update(updated_params)
model.load_state_dict(new_params)
```

**Comment by karandwivedi42** [comment]
Comparing to [`load_state_dict`](https://github.com/pytorch/pytorch/blob/2c038f20748c588c9e98f3d549036237ab280ff7/torch/nn/modules/module.py#L337)  which uses `copy_`, this workaround doesn't. Does that matter?

For fine-tuning it becomes:
```python
updated_params = torch.load(pretrained)
updated_params.pop('fc.weight', None)
updated_params.pop('fc.bias', None)
new_params = model.state_dict()
new_params.update(updated_params)
model.load_state_dict(new_params)
```

`ignored_keys` can just ignore the given keys from both loaded `state_dict` and `own_state` like:

```python
    def load_state_dict(self, state_dict, ignored_keys=[]):
        """Copies parameters and buffers from :attr:`state_dict` into
        this module and its descendants. The keys of :attr:`state_dict` must
        exactly match the keys returned by this module's :func:`state_dict()`
        function.
        Arguments:
            state_dict (dict): A dict containing parameters and
                persistent buffers.
        """
        own_state = self.state_dict()
        for name, param in state_dict.items():
            if name not in ignored_keys:
                if name not in own_state:
                    raise KeyError('unexpected key "{}" in state_dict'
                                   .format(name))
                if isinstance(param, Parameter):
                    # backwards compatibility for serialized parameters
                    param = param.data
                own_state[name].copy_(param)

        missing = set(own_state.keys()) - set(state_dict.keys()) - set(ignored_keys)
        if len(missing) > 0:
            raise KeyError('missing keys in state_dict: "{}"'.format(missing))
```

**Comment by colesbury** [comment]
I don't think we should add `ignored_keys`. If you know the list of extra keys, it's easy to delete them from the state_dict before you call `load_state_dict`

**Comment by soumith** [comment]
it feels like something you'd do very often for fine-tuning, and seems like adding the option would benefit from reducing the verbosity of the code (In such a common use-case)

**Comment by cedias** [comment]
It's also annoying if there are some buffers you don't want to save/reload. Being able to ignore them explicitly or being able to load an incomplete state dictionnary would be much less verbose.

**Comment by alykhantejani** [comment]
@apaszke I think this can now be closed as #3159 is merged
